Library/Homebrew/formula_versions.rb

```diff
@@ -15,56 +15,64 @@ class FormulaVersions
     ErrorDuringExecution, LoadError, MethodDeprecatedError
   ].freeze
 
-  MAX_VERSIONS_DEPTH = 2
-
-  attr_reader :name, :path, :repository, :entry_name
-
+  sig { params(formula: Formula).void }
   def initialize(formula)
     @name = formula.name
     @path = formula.path
-    @repository = formula.tap.path
-    @entry_name = @path.relative_path_from(repository).to_s
-    @current_formula = formula
+    @repository = T.must(formula.tap).path
+    @relative_path = @path.relative_path_from(repository).to_s
+    # Also look at e.g. older homebrew-core paths before sharding.
+    if (match = @relative_path.match(%r{^(HomebrewFormula|Formula)/([a-z]|lib)/(.+)}))
+      @old_relative_path = "#{match[1]}/#{match[3]}"
+    end
     @formula_at_revision = {}
   end
 
   def rev_list(branch)
     repository.cd do
       rev_list_cmd = ["git", "rev-list", "--abbrev-commit", "--remove-empty"]
-      rev_list_cmd << "--first-parent" if repository != CoreTap.instance.path
-      Utils.popen_read(*rev_list_cmd, branch, "--", entry_name) do |io|
-        yield io.readline.chomp until io.eof?
+      [relative_path, old_relative_path].compact.each do |entry|
+        Utils.popen_read(*rev_list_cmd, branch, "--", entry) do |io|
+          yield [io.readline.chomp, entry] until io.eof?
+        end
```

Comment:
I _think_ you can just pass both paths to `git rev-list`? Probably will be faster. Unless you need to pass them separately in order to correctly process the output? But it doesn't seem like it.

Reply:
@carlocab You can do both at once but it only returns a Git SHA and then you don't know what `entry` the SHA refers to for `git cat-file` later (which spits out errors to stdout if paths don't exist).

It's a bit slower but not dramatically:

```console
$ hyperfine "git rev-list --abbrev-commit --remove-empty master -- Formula/a/ack.rb Formula/ack.rb"
Benchmark 1: git rev-list --abbrev-commit --remove-empty master -- Formula/a/ack.rb Formula/ack.rb
  Time (mean ± σ):      7.874 s ±  0.164 s    [User: 7.459 s, System: 0.396 s]
  Range (min … max):    7.596 s …  8.032 s    10 runs
$ hyperfine "git rev-list --abbrev-commit --remove-empty master -- Formula/a/ack.rb; git rev-list --abbrev-commit --remove-empty master -- Formula/ack.rb" 
Benchmark 1: git rev-list --abbrev-commit --remove-empty master -- Formula/a/ack.rb; git rev-list --abbrev-commit --remove-empty master -- Formula/ack.rb
  Time (mean ± σ):      8.310 s ±  0.572 s    [User: 7.475 s, System: 0.806 s]
  Range (min … max):    7.769 s …  9.825 s    10 runs
```